src/main/java/tigase/server/rtbl/RTBLRepository.java

```diff
@@ -0,0 +1,459 @@
+				} catch (TigaseDBException ex) {
+					logger.log(Level.WARNING, "failed to save updated RTBL " + key + ", adding " + id + " failed", ex);
+				}
+			}
+			case remove -> {
+				try {
+					userRepository.removeData(repoUser, key.getSubnode(), id);
+				} catch (TigaseDBException ex) {
+					logger.log(Level.WARNING, "failed to save updated RTBL " + key + ", removing " + id + " failed", ex);
+				}
+			}
+		}
+	}
+	
+	private RTBL loadList(Key key) throws TigaseDBException {
+		if (userRepository.userExists(repoUser)) {
+			try {
+				Map<String, String> data = userRepository.getDataMap(repoUser, key.getSubnode());
+				BareJID jid = BareJID.bareJIDInstanceNS(data.remove("jid"));
+				String node = data.remove("node");
+				String hash = data.remove("hash");
+				if (jid == null || hash == null || node == null) {
+					return null;
+				}
+				return new RTBL(jid, node, hash, new CopyOnWriteArraySet<>(data.keySet()));
+			} catch (TigaseDBException ex) {
+				logger.log(Level.WARNING, "could not load RTBL for " + key, ex);
+			}
+		}
+		return null;
+	}
+
+	protected void reload() {
+		try {
+			if (userRepository.userExists(repoUser)) {
+				String[] subnodes = userRepository.getSubnodes(repoUser);
+				if (subnodes != null) {
+					for (String subnode : subnodes) {
+						Key key = Key.parse(subnode);
+						try {
+							if (key != null) {
+								RTBL rtbl = loadList(key);
+								if (rtbl != null) {
+									cache.put(key, rtbl);
+								} else {
+									cache.remove(key);
+								}
+							}
+						} catch (TigaseDBException ex) {
+							logger.log(Level.WARNING, "could not load RTBL " + key, ex);
+						}
+					}
+				}
+			}
+		} catch (TigaseDBException ex) {
+			logger.log(Level.WARNING, "could not load list of RTBLs", ex);
+		}
+	}
+
+	public static class Key {
+		private final BareJID jid;
+		private final String node;
+
+		public Key(BareJID jid, String node) {
+			this.jid = jid;
+			this.node = node;
+		}
+
+		public String getNode() {
+			return node;
+		}
+
+		public BareJID getJid() {
+			return jid;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (!(o instanceof Key key)) {
+				return false;
+			}
+			return Objects.equals(jid, key.jid) && Objects.equals(node, key.node);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(jid, node);
+		}
+
+		public String getSubnode() {
+			return jid.toString() + "|" + node.replaceAll("/", "|");
+		}
+
+		public static Key parse(String data) {
+			String[] parts = data.split("\\|");
+			if (parts != null && parts.length > 1) {
+				return new Key(BareJID.bareJIDInstanceNS(parts[0]),
+							   Arrays.stream(parts).skip(1).collect(Collectors.joining("|")));
+			}
+			return null;
+		}
+
+		@Override
+		public String toString() {
+			return "jid: " + jid + ", node: " + node;
+		}
+	}
+
+	public static abstract class RTBLEvent implements Serializable, EventBusEvent {
```

Comment:
Hmm... I know it's abstract and rest of the events extend it but maybe convert them to `record`s as well?

Reply:
It is impossible to use this "shiny" record, as it requires empty constructor and values may be set using setters. AFAIR, it is not possible with `record` and I'm not sure that record can be extended...

Reply:
No, it's not possible to extend records nor to have empty constructor / setters.

There's no need to be sarcastic and snarky with _"shiny"_. Simple "events require empty constructor and setters for serialisation and deserialization" would be enough.

Reply:
Unfortunately the problem with review on GH. It was mentioned in case of Key (small class) and events, but there were a couple of events so some repetitiveness to mark places. I could probably mention all the places in single comment but it _felt_ better to mark them with comment.

Shorter code is usually better. Using `record` right now let's you skip having to implement hashCode, equals, toString, etc. It's immutable so there's a benefit as well, an in the future they could possibly be value/primitive class, that have additional benefits regarding memory usage and allocation.

Obviously in the future things may change and require alterations, but it should be relatively easy to convert record to a full-blown class. My suggestion was more in the vain of _KISS_ - _keep it simple_